pulumi/infra/kafka.py

```diff
@@ -1,13 +1,515 @@
+        return CredentialOutput(
+            service_account_id=json_["service_account_id"],
+            api_key=json_["api_key"],
+            api_secret=json_["api_secret"],
+        )
+
+
+@dataclasses.dataclass
+class EnvironmentOutput:
+    environment_id: str
+    bootstrap_servers: str
+    environment_credentials: CredentialOutput
+    service_credentials: Mapping[str, CredentialOutput]
+
+    def get_service_credentials(self, service_name: str) -> CredentialOutput:
+        if service_name in self.service_credentials:
+            return self.service_credentials[service_name]
+        else:
+            raise KeyError(f"{service_name} does not exist")
+
+    @staticmethod
+    def from_json(json_: Mapping[str, Any]) -> EnvironmentOutput:
+        return EnvironmentOutput(
+            environment_id=json_["environment_id"],
+            bootstrap_servers=json_["bootstrap_servers"],
+            environment_credentials=CredentialOutput.from_json(
+                json_["environment_credentials"]
+            ),
+            service_credentials={
+                k: CredentialOutput.from_json(v)
+                for k, v in json_["service_credentials"].items()
+            },
+        )
+
+
+@dataclasses.dataclass
+class ConfluentOutput:
+    environments: Mapping[str, EnvironmentOutput]
+
+    def get_environment(self, environment_name: str) -> EnvironmentOutput:
+        if environment_name in self.environments:
+            return self.environments[environment_name]
+        else:
+            raise KeyError(f"{environment_name} does not exist")
+
+    @staticmethod
+    def from_json(json_: pulumi.Output[Mapping[str, Any]]) -> pulumi.Output[ConfluentOutput]:
+        return json_.apply(
+            lambda j: ConfluentOutput(
+                environments={
+                    k: EnvironmentOutput.from_json(v) for k, v in j.items()
+                }
+            )
+        )
 
 
 class Kafka(pulumi.ComponentResource):
-    def __init__(self, name: str, opts: Optional[pulumi.ResourceOptions] = None):
+    def __init__(
+        self,
+        name: str,
+        confluent_environment_name: str,
+        opts: Optional[pulumi.ResourceOptions] = None,
+    ):
         super().__init__("grapl:Kafka", name=name, props=None, opts=opts)
-        metrics_topic = kafka.Topic(
-            "metrics-topic", name="metrics", replication_factor=1, partitions=1
+
+        confluent_stack_output = StackReference(
+            "grapl/ccloud-bootstrap/ccloud-bootstrap"
+        ).get_output("confluent")
+
+        assert confluent_stack_output is not None
+
+        confluent_environment = ConfluentOutput.from_json(
+            cast(pulumi.Output[Mapping[str, Any]], confluent_stack_output)
+        ).apply(lambda o: o.get_environment(confluent_environment_name))
+
+        if LOCAL_GRAPL:
+            provider = kafka.Provider(
+                "kafka-provider",
+                bootstrap_servers=[os.environ["KAFKA_ENDPOINT"]],
```

Comment:
So, I'd bet good money that environ has no KAFKA_ENDPOINT right now.
I suspect what you'll want is
f"localhost:{os.environ["KAFKA_BROKER_PORT_FOR_HOST_OS"]}"

The reasoning here is complicated - I'd suggest line 226 of nomad/local/grapl-local-infra.nomad